org.civicrm.reqangular/src/common/services/file-upload.js

```diff
@@ -0,0 +1,81 @@
+define([
+  'common/modules/services',
+], function (module) {
+  'use strict';
+
+  module.factory('FileUploadService', ['$q', 'FileUploader',
+    function ($q, FileUploader) {
+      var fileUploader;
+
+      return {
+
+        /**
+         * Sets up the file uploader service
+         *
+         * @param {Object} customSettings can have keys like url for file upload path,
+         * formData for updating form data, queueLimit to limit number of files that can be uploaded at onCompleteAll
+         *
+         * @return {Object} an instance of FileUploader
+         **/
+        uploader: function (customSettings) {
+          var uploaderSettings = {};
+
+          uploaderSettings.url = 'url' in customSettings ? customSettings.url : '/civicrm/ajax/attachment';
+          uploaderSettings.formData = [{
+            entity_table: customSettings.entityTable,
+            crm_attachment_token: customSettings.crmAttachmentToken
+          }];
+
+          if (customSettings.queueLimit && typeof customSettings.queueLimit === 'number') {
+            uploaderSettings.queueLimit = customSettings.queueLimit;
+          }
+
+          fileUploader = new FileUploader(uploaderSettings);
+          return fileUploader;
+        },
+
+        /**
+         * Uploads all files in queue updating with additional form data
+         *
+         * @param {Object} additionalFormData that has keys like entityID or description
+         **/
+        uploadAll: function (additionalFormData) {
+          var deferred = $q.defer(),
+            results = [];
+
+          fileUploader.onBeforeUploadItem = function (item) {
```

Comment:
* FileUploader callback that gets fired before each file item gets uploaded, useful 
          * to insert additional form params like entityId which are not available 
          * at the time of uploader creation.